plugin/pkg/auth/authenticator/password/keystone/keystone.go

```diff
@@ -45,18 +50,74 @@ func (keystoneAuthenticator *KeystoneAuthenticator) AuthenticatePassword(usernam
 		glog.Info("Failed: Starting openstack authenticate client")
 		return nil, false, errors.New("Failed to authenticate")
 	}
-
+//FIXME user_id, not user_name...
 	return &user.DefaultInfo{Name: username}, true, nil
 }
 
-// NewKeystoneAuthenticator returns a password authenticator that validates credentials using openstack keystone
-func NewKeystoneAuthenticator(authURL string) (*KeystoneAuthenticator, error) {
-	if !strings.HasPrefix(authURL, "https") {
-		return nil, errors.New("Auth URL should be secure and start with https")
+func (keystoneAuthenticator *KeystoneAuthenticator) AuthenticateToken(value string) (user.Info, bool, error) {
+//FIXME kfox1111: To enhance performance, cache admin token/client.
+	cfg, provider, err := loadConfig(keystoneAuthenticator.configPath)
+	if err != nil {
+		return nil, false, errors.New("Internal error getting the Keystone provider")
+	}
+
+	client, err := openstack.NewIdentityAdminV3(provider, gophercloud.EndpointOpts{
+		Region: cfg.Global.Region,
+	})
+	if err != nil {
+		return nil, false, err
+	}
+	response, err := client.Request("GET", client.ServiceURL("auth", "tokens"), gophercloud.RequestOpts{
+	        MoreHeaders: map[string]string{"X-Subject-Token": value},
+	        OkCodes:     []int{200},
+	})
+	if err != nil {
+		return nil, false, err
+	}
+	defer response.Body.Close()
+	bodyBytes, err := ioutil.ReadAll(response.Body)
+	if err != nil {
+		glog.V(3).Infof("Cannot get HTTP response body from keystone token validate: %v", err)
+		return nil, false, err
+	}
+	obj := struct{ Token struct { User struct { Id string }}}{}
+	err = json.Unmarshal(bodyBytes, &obj)
+	if err != nil {
+		return nil, false, err
+	}
+	return &user.DefaultInfo{Name: obj.Token.User.Id}, response.StatusCode == 200, nil
```

Comment:
I think that may be able to be handled by just adding an additional Extra key that tags the auth as having come from the keystone auth plugin. Then the authz plugin would only work if the flag is in Extra?

That should allow a service account through.

I think the service account thing could be abused in a multi tenant scenario though.

I'm going to work on only the single tenant use case for now, since that's pretty safe. We probably should talk at the sig-auth meeting about what multitenant might look like. It may be much more invasive, and something k8s doesn't want to support at all. But I think would be a great feature.